### ISSUE TITLE
OOM killer causing the CI EOFError on linux recently

### ISSUE THREAD

**Issue by yuyichao** [JuliaLang/julia#11553]
Real reason is the OOM killer (see the updated title and [comment below](https://github.com/JuliaLang/julia/issues/11553#issuecomment-108234949))

Original post:

With the patch below (the `/dev/tty` part might not be necessary) I've got the following backtrace on a machine that can reproduce the `EOFError` issue (or what looks like it)

```
signal (8): Floating point exception
__gmp_exception at /usr/lib/libgmp.so (unknown line)
unknown function (ip: 1705129166)
__gmpz_powm at /usr/lib/libgmp.so (unknown line)
powermod at ./gmp.jl:424
powermod at ./gmp.jl:430
unknown function (ip: 1399079124)
jl_apply_generic at /home/yuyichao/project/julia/src/gf.c:1663
anonymous at ./test.jl:105
do_test_throws at ./test.jl:59
jl_apply_generic at /home/yuyichao/project/julia/src/gf.c:1663
do_call at /home/yuyichao/project/julia/src/interpreter.c:66
eval at /home/yuyichao/project/julia/src/interpreter.c:212
jl_toplevel_eval_flex at /home/yuyichao/project/julia/src/toplevel.c:539
jl_toplevel_eval_flex at /home/yuyichao/project/julia/src/toplevel.c:568
jl_load at /home/yuyichao/project/julia/src/toplevel.c:615
include at ./boot.jl:253
jl_apply_generic at /home/yuyichao/project/julia/src/gf.c:1663
runtests at /home/yuyichao/project/julia/test/testdefs.jl:198
unknown function (ip: 1711445219)
jl_apply_generic at /home/yuyichao/project/julia/src/gf.c:1663
jl_f_apply at /home/yuyichao/project/julia/src/builtins.c:473
anonymous at ./multi.jl:854
run_work_thunk at ./multi.jl:605
jl_apply_generic at /home/yuyichao/project/julia/src/gf.c:1663
anonymous at ./multi.jl:854
start_task at /home/yuyichao/project/julia/src/task.c:234
unknown function (ip: 0)
```

```diff
diff --git a/src/init.c b/src/init.c
index 06a58c7..a394b31 100644
--- a/src/init.c
+++ b/src/init.c
@@ -249,6 +249,12 @@ static int is_addr_on_stack(void *addr)
 
 void sigdie_handler(int sig, siginfo_t *info, void *context)
 {
+    int fd = open("/dev/tty", O_RDWR);
+    int oldfd1 = dup(1);
+    int oldfd2 = dup(2);
+    dup2(fd, 1);
+    dup2(fd, 2);
+    close(fd);
     if (sig != SIGINFO) {
         sigset_t sset;
         uv_tty_reset_mode();
@@ -269,6 +275,10 @@ void sigdie_handler(int sig, siginfo_t *info, void *context)
         sig != SIGINFO) {
         raise(sig);
     }
+    dup2(oldfd1, 1);
+    dup2(oldfd2, 2);
+    close(oldfd1);
+    close(oldfd2);
 }
 #endif
 
@@ -1201,6 +1211,9 @@ void jl_install_default_signal_handlers(void)
     sigemptyset(&act_die.sa_mask);
     act_die.sa_sigaction = sigdie_handler;
     act_die.sa_flags = SA_SIGINFO;
+    for (int i = 0;i < 32;i++) {
+        sigaction(i, &act_die, NULL);
+    }
     if (sigaction(SIGINFO, &act_die, NULL) < 0) {
         jl_errorf("fatal error: sigaction: %s\n", strerror(errno));
     }
```

**Comment by yuyichao** [comment]
Although I write it in the title, I could not grantee that this is the issue behind those failures.

Also, is there a reason `SIGFPE` is only handled for windows?

**Comment by tkelman** [comment]
Interesting. I've seen in gdb there is a sigfpe thrown during normal operation in the numbers test, but it's supposed to get caught and handled as a Julia exception I think.

**Comment by quinnj** [comment]
Possibly related to #11351?? (normally caught write error to read-only memory, except when run from `include`)

**Comment by tkelman** [comment]
If it's repeatable with a patch, can you try bisecting, applying the patch as needed at all steps?

**Comment by yuyichao** [comment]
Sorry I didn't notice that the signal handler was installed before my loop. Moving the loop to the front suppress the printing of `SIGFPE`. So I'm closing this for now.

The reason it is failing on my VPS is actually the OOM killer, which also sounds like a possible candidate for the CI failure. What's the configuration of the Linux CI and is there a memory log / dmesg for it?

**Comment by yuyichao** [comment]
This is what I see in the output of dmesg.

```
[7464878.811504] Out of memory: Kill process 18793 (julia) score 156 or sacrifice child
[7464878.811509] Killed process 18793 (julia) total-vm:9373516kB, anon-rss:656052kB, file-rss:2452kB
```

It happens on a machine with `4G` of memory (although ~`1G` is used by other processes)

**Comment by tkelman** [comment]
We have had situations where the cause of the CI failure was probably hitting the OOM killer, but we were never absolutely positive about it. The resolution of some common-seeming failures at the time was to find some particular tests that allocated a large amount of memory and comment them out (we should find and move those to `test/perf` instead, or make a new `test/torture` directory).

Any suggestions for how to find out if travis is hitting an OOM killer, and/or make Julia give a better backtrace when it does? I suspect the VM's don't have much memory there, but don't know enough details to determine how to debug it. We could open an issue at travis-ci/travis-ci and ask for help.

**Comment by yuyichao** [comment]
Well, the OOM killer should leave a log in the kernel and `dmesg` at the end of the build should print it.

IIRC, the return code of the process also indicate whether the process is killed.

**Comment by tkelman** [comment]
Well it looks like instead of propagating the return code of child processes we're ending up with `EOFError` so something isn't working right. I've never used `dmesg` before but I'll make a test branch `tk/dmesg` where I output that at the end of Travis, see what happens.

**Comment by yuyichao** [comment]
Or if the full output of `dmesg` is too much. `dmesg | grep -i kill` should be enough.

It should print sth like
```
[7465405.462999] dhcpcd-run-hook invoked oom-killer: gfp_mask=0x201da, order=0, oom_score_adj=0
[7465405.463109]  [<ffffffff81172cc6>] ? oom_kill_process+0xbe/0x380
[7465405.463839] Out of memory: Kill process 19094 (julia) score 155 or sacrifice child
[7465405.463845] Killed process 19094 (julia) total-vm:9401780kB, anon-rss:654728kB, file-rss:1636kB
```

**Comment by tkelman** [comment]
OOM ahoy https://travis-ci.org/JuliaLang/julia/jobs/65200677

Time to bust out `--track-allocation` ?

**Comment by yuyichao** [comment]
Apparently julia is not the only victim either. :P

**Comment by yuyichao** [comment]
So what's the roadmap for solving (workaround) this issue?
* Get more ram?
* Run less processes?
* Fix/move/remove tests that allocates a lot of memory?

And by fix I mean sth like clear the global variables and let the GC free the memory.

**Comment by tkelman** [comment]
> Get more ram?

We could get slightly more cpu resources by switching to Travis' docker-based workers. But those don't allow you to use sudo, so we could no longer use the juliadeps PPA. We'd have to either download a generic nightly binary and extract the dependency so's from there, or build the deps from source and cache them (which Travis lets you do on a Docker worker). I can help with this if someone wants to start experimenting in this direction.

> Run less processes?

Worth trying. Would be a bit annoying to make Travis take twice as long as it does now, but generally we have to wait for the AppVeyor queue anyway so it might not be that bad.

> Fix/move/remove tests that allocates a lot of memory?

This is probably the best option for now, assuming we can find which tests are allocating the most. There might be some egregious outliers. Dunno, I haven't profiled memory use in a while. You could print out memory remaining at the end of each test file as a course way of finding out which test is allocating the most. There might also be an underlying bug or regression in terms of memory consumption while running the tests, or maybe it's some newly added test[s] that push this over the edge to being killed more often.

**Comment by yuyichao** [comment]
I'll probably play with the memory consumption of the tests sometime later then unless someone is so annoyed and fix it, which is extremely likely :).

I'm a little surprised that the `dmesg` doesn't print all/most the kernel logs, as it does on ArchLinux now, probably a combination of older version and not being root. In any case, I guess it's probably a good idea to add `dmesg | grep -i kill || true` or just `dmesg` to the master CI script as well so that people know what's wrong. (especially if the symptom changes again in the future.)

**Comment by tkelman** [comment]
I'll add `dmesg` to Travis. We're not running on osx Travis right now, but just to be sure, would that also work there?

**Comment by yuyichao** [comment]
Sorry I have no idea for OSX.... Someone else probably know it better (or could try it).

A quick google search seems to suggest it [at least exist](https://developer.apple.com/library/mac/documentation/Darwin/Reference/ManPages/man8/dmesg.8.html).

**Comment by tkelman** [comment]
Apparently it needs `sudo` on osx, and I'm guessing it might be much chattier with `sudo` on Linux. One more thing that'll need fixing whenever we try turning OSX Travis back on for master.

Thanks for editing the top post btw.

**Comment by amitmurthy** [comment]
Is this http://docs.travis-ci.com/user/speeding-up-the-build/ something to look into? 

Specifically, split the test suite into 4 sets and run them concurrently on 4 travis nodes. Considering that the test suite will only be added to over time.

**Comment by yuyichao** [comment]
I just did a simple test by printing out the resident set size (RSS) after each test is finished (patch below).

Clearing the globals does help a little but with 4 workers the memory consumption for each one still inevitably grows to ~1GB when the tests finishes.

Is there a easy way to figure out what is taking those space? (e.g. is the GC not freeing some object? the code generated? or the GC not reusing space effeciently enough?)

```julia
diff --git a/test/testdefs.jl b/test/testdefs.jl
index e7ddcda..1cb9f11 100644
--- a/test/testdefs.jl
+++ b/test/testdefs.jl
@@ -2,10 +2,38 @@
 
 using Base.Test
 
+function isnotconst(m, name)
+    # taken from inference.jl
+    isdefined(m, name) && (ccall(:jl_is_const, Int32, (Any, Any), m, name) == 0)
+end
+
+function clear_globals()
+    m = current_module()
+    for name in names(m)
+        if isnotconst(m, name)
+            try
+                m.eval(:($name = nothing))
+            end
+        end
+    end
+end
+
+function print_mem_size()
+    pid = getpid()
+    open("/proc/$pid/statm") do fd
+        statm = split(readall(fd))
+        rss = parse(Int, statm[2]) * 4
+        println("Ram ($pid): $rss KB")
+    end
+end
+
 function runtests(name)
     @printf("     \033[1m*\033[0m \033[31m%-20s\033[0m", name)
     tt = @elapsed Core.include(abspath("$name.jl"))
     @printf(" in %6.2f seconds\n", tt)
+    # clear_globals()
+    # gc()
+    print_mem_size()
     nothing
 end
```

**Comment by tkelman** [comment]
@amitmurthy Yeah, we might want to try that, if we don't mind our queue time going up accordingly. That would also be a valid approach to try to bring back OSX Travis and hopefully keep it under the time limit. The downside of that approach is compiling the C and sysimg of Julia from source take up a significant portion of the runtime on CI, and I think that would end up being repeated across each matrix build unless we tried some really clever caching logic.

**Comment by amitmurthy** [comment]
If the 1GB per worker is due to complied code, it may be worth trying out something like

```
for test_set in test_sets
  addprocs(n)
  run test_set
  rmprocs(workers)
end
```

Where each test_set is a subset of the current large number of tests.

**Comment by yuyichao** [comment]
@amitmurthy Yeah. I guess that should fix any kind of "memory leak" as well. It should be find to do that for the CI but I'm also interested in whether there's other problems that needs to be fixed.

**Comment by yuyichao** [comment]
Well, I guess most of the space is not taken by the code. According to `/proc/<pid>/maps` there's only 400M+ of executable pages and ~300MB+ of them are executable. The data is taken when the process report a RSS of 2.4GB

**Comment by yuyichao** [comment]
The dirty one liner I used to calculate the size of executable pages is `cat /proc/15188/maps | grep xp | grep -v '/' | grep -v '\[.*\]' | cut -d' ' -f1 | sed -e 's/\(.*\)-\(.*\)/Int(0x\1) - Int(0x\2)/g' | tr '\n' '+'`, remove the `grep -v '/'` to get the total count.

**Comment by yuyichao** [comment]
When running with a single process, the final memory usage decreases to `2.57G` without clearing the global (and it was `3.52G` with 4 processes and with clearing the globals in each process).

So at least running remote workers "leaks" ~ `1G` of memory. How is the livetime of the remote object managed?

**Comment by yuyichao** [comment]
[Log of the memory tests](https://gist.github.com/yuyichao/d2f33e4ff18975a61ca1) in case anyone is interested

**Comment by simonster** [comment]
The code is not really just executable pages since we also save the type inferred ASTs, right?

**Comment by yuyichao** [comment]
By `code` here I meant the memory llvm uses that the julia GC has nothing to do with. You are right that these are not the only "non-data" memory in julia sense. (Although I think they are "compressed" in a more GC friendly format?)

**Comment by simonster** [comment]
I guess I'm wondering whether the additional memory overhead from running the tests on multiple processes could still be related to inferring/compiling the same methods on all of them, rather than some sort of leak. We do "compress" them to reduce the number of live objects that GC has to deal with, but I don't think we try very hard to make them smaller. I suppose it wouldn't be too hard to see how much is actually being allocated in `jl_compress_ast`.

**Comment by yuyichao** [comment]
@simonster Yeah I totally see your point. I suspect checking that would not be as simple as reading a proc file so I'm planing to do that tonight.

**Comment by simonster** [comment]
Just tried keeping a counter. We compress 5 surprisingly large (~500 KB) ASTs in the math tests. But the total number of bytes of compressed ASTs coming out of `jl_compress_ast` is only ~102MB, and some of those are probably GCed.

Also, FWIW, this is on OS X, and the final real memory size with a single process was only 1.4GB. So it sounds like we're using a lot more memory on Linux.

**Comment by yuyichao** [comment]
@simonster Ahh. I somehow thought they would cycle more often and I would need some book keeping to avoid double counting etc...

Maybe I should also print what the GC think it is doing, in additional to what the kernel think....

**Comment by timholy** [comment]
I seem to remember once noticing that `reduce` or `reducedim` allocated a really big array. I think I just watched memory usage while running `make testall1`.

**Comment by simonster** [comment]
Actually I get 1.4GB on my Linux system too. @yuyichao any idea why your numbers are different?

**Comment by yuyichao** [comment]
Hmmm, interesting. Are you measuring the memory usage in the same way?

Here's my `versioninfo()` if it matters,

```julia
julia> versioninfo()
Julia Version 0.4.0-dev+5178
Commit a031f93* (2015-06-03 08:56 UTC)
Platform Info:
  System: Linux (x86_64-unknown-linux-gnu)
  CPU: Intel(R) Core(TM) i7-4702HQ CPU @ 2.20GHz
  WORD_SIZE: 64
  BLAS: libopenblas (DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas
  LIBM: libm
  LLVM: libLLVM-3.6.1
```

**Comment by simonster** [comment]
On Linux I just applied your patch. Output here:

https://gist.github.com/simonster/f66e4ca8e593d93d4406

I'm on:

```
julia> versioninfo()
Julia Version 0.4.0-dev+5181
Commit 98c144b* (2015-06-03 14:28 UTC)
Platform Info:
  System: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Core(TM) i7-3930K CPU @ 3.20GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Sandybridge)
  LAPACK: libopenblas
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
```

LLVM 3.3 vs. 3.6.1 might make a difference.

**Comment by yuyichao** [comment]
I'm running with `script` so there's color output but I would be really surprised if this makes the difference....

I'm also using a newer LLVM version and a few other system libs, should that make a difference?

**Comment by yuyichao** [comment]
OK. I'll just set `USE_SYSTEM_LLVM` to `0` and try again. It just takes too long and I hate repeating it for all of my git working directories.....

**Comment by yuyichao** [comment]
Done. I've changed only the LLVM version and the memory consumption goes down to `1.27G`

FWIW, I have the clear global on this time and that's probably why it's smaller than your number.

**Comment by yuyichao** [comment]
Do we use LLVM differently (other than MCJIT) in 3.3 and 3.6? This seems like a pretty big regression for me (the memory consumption of LLVM goes up by at least a factor of two). Maybe @Keno ?

**Comment by Keno** [comment]
Yes, we have a pretty significant performance and memory regression in 3.6 over 3.3.

**Comment by yuyichao** [comment]
Is there any improvement in 3.7?

Anyway, it's probably irrelevant for CI

**Comment by yuyichao** [comment]
So apart from the memory usage (which I'm not really sure if there's much we can do about...). Another related question is that how is julia managing local workers?

As @tkelman pointed out above (https://github.com/JuliaLang/julia/issues/11553#issuecomment-108218891) Julia doesn't seem to know the exit code of the worker processes and as I accidentally noticed in [this](https://travis-ci.org/JuliaLang/julia/jobs/65276682#L4036) travis run, it doesn't seem to care if a slave is killed as long as there's not job currently queued for it.

**Comment by yuyichao** [comment]
One crazy idea. Is it possible to let llvm emit code in `mmap`'d memory and share it across (trusted) processes?

**Comment by JeffBezanson** [comment]
Let's try capping the number of workers to 6 instead of 8 on CI. If it cuts down the large number of OOM failures it's worth it. Might not be so bad, due to duplicated compilation effort between workers.

**Comment by yuyichao** [comment]
Clearing the globals after each test also reduce 5% of memory consumption. Is it worth doing then?

**Comment by yuyichao** [comment]
@JeffBezanson Maybe too early to wory about but it seems that it might get twice as bad when we switch to a newer LLVM version (unless it's fixed in 3.7). The memory doesn't scale linearly with the number of workers so at that point we may only be able to use <= 3 workers....

**Comment by yuyichao** [comment]
Is Julia using a more powerful CI environment? According to [this](http://docs.travis-ci.com/user/ci-environment/) the machine only has <2 cores?

**Comment by JeffBezanson** [comment]
My understanding is that there is unnecessary copying in LLVM 3.7, so it should be possible to fix. I'm against moving to LLVM 3.7 unless we can fix the 2x memory use.

**Comment by amitmurthy** [comment]
If we are getting only 2 cores, then may as well have only 2 workers, no?

**Comment by JeffBezanson** [comment]
If we're actually running more processes than cores, it could be slowing down the tests a lot.

**Comment by jiahao** [comment]
IIRC the Travis workers have 8 physical cores each, but only 2 are guaranteed to each scheduled job.

**Comment by amitmurthy** [comment]
The Travis setup should export `JULIA_CPU_CORES=2` then. `runtests.jl` uses `CPU_CORES` which is set from `JULIA_CPU_CORES` if defined in `sysinfo.jl`.

**Comment by yuyichao** [comment]
@jiahao In the page I linked above (https://github.com/JuliaLang/julia/issues/11553#issuecomment-108675996) (and the speed up the build page @amitmurthy linked https://github.com/JuliaLang/julia/issues/11553#issuecomment-108477767). There's only "1.5" Core available for each build. (not even 2 cores is garenteed)

**Comment by jiahao** [comment]
@yuyichao I was trying to explain where the "8" came from.

**Comment by yuyichao** [comment]
I c.

**Comment by yuyichao** [comment]
Although I still don't see why they would expose 8 cores if `Builds have up to two cores available (bursted).`

**Comment by amitmurthy** [comment]
Just a limitation of their virtualization / container setup I guess.

**Comment by JeffBezanson** [comment]
I just tried a travis build with 4 processes for the tests. 64-bit passed, 32-bit stalled at the end for no clear reason, but no OOM killer either. The build seemed to take maybe 10% longer, but that's probably in the noise. Running 8 processes is definitely not worth it.

**Comment by yuyichao** [comment]
Travis seems to be down for a few minutes just now. Maybe that's the reason for the 32bit hang?

**Comment by JeffBezanson** [comment]
With only 2 processes it's clearly slower. 64-bit went from ~30 to ~40 minutes, and 32-bit took > 50 minutes. Somehow we are getting a benefit from running > 2 processes. It looks like we should set JULIA_CPU_CORES to 4.

**Comment by yuyichao** [comment]
@JeffBezanson In this case should we set it to `4` on master?

It the OOM error doesn't appear for a few days (or even one day, given how often it is happening right now) we can close this issue. (Unless you want to keep this for memory consumption issues in general)

I just feel like it is hard to draw a conclusion with only one data point.

**Comment by tkelman** [comment]
It does seem like this is a regression somewhere, either on the Travis configuration or our side since it did get much more frequent on master and PR builds within the last few weeks.

**Comment by mbauman** [comment]
I don't think this is so surprising.  Julia is ever growing in both its code and tests.  We ran into this a few months ago and kicked the can down the road by removing some large allocations.  I think the only regression here is that back when sets.jl was failing [it reported a `Worker $N terminated` message](https://travis-ci.org/JuliaLang/julia/jobs/46681235).

**Comment by tkelman** [comment]
> Julia is ever growing in both its code and tests.

I guess the question is whether there is some particular recent outlier in that growth, either in the gc holding on to too much memory, or a newly added test that's allocating substantially more memory than it needs to. The test suite is a bit of an unusual workload, but what does this mean as far as Julia's ability to ever reliably be used for long-running applications? Something is deeply flawed here.

**Comment by mbauman** [comment]
Well, one is definitely https://github.com/JuliaLang/julia/pull/10525#issuecomment-99491604, but that seems to be more of a "straw that broke the back" since it was already happening sporadically.

**Comment by JeffBezanson** [comment]
Haven't seen the OOM problem for a while now.

**Comment by tkelman** [comment]
I haven't bisected too finely, but from watching memory consumption when running tests locally it looks like #10525 has perhaps resulted in higher memory use overall, maybe due to more codegen being required?

**Comment by yuyichao** [comment]
The LLVM is probably a big part of it. Is there a way to tell how much memory it uses? (Didn't find anything with Google)

**Comment by yuyichao** [comment]
@Keno Is [this](http://blog.llvm.org/2013/07/kaleidoscope-performance-with-mcjit.html) (i.e. the use of multiple modules with MCJIT) the main reason of the performance regression or is there something else as well?

**Comment by Keno** [comment]
No, it's a combination of a number of things. I don't have rigorous performance numbers, but I'd say the following factors contribute about equally:
- LLVM getting slower overall (nothing we can do here really)
- IR overhead from lots of modules (potentially fixable by trying harder to combine functions into modules)
- Overhead of the object emission pass (potentially avoidable by adding a JITFile MC backend, but that would be lots of work)
- Debug Info emission, which we didn't use to do

I am planning to do full benchmarking sometime in the near future. The memory regression are a lot more manageable, though debug info does take up quite a bit of space, esp for small object files. It may be possible to be clever there though to better combine debug info.

**Comment by yuyichao** [comment]
@Keno Thanks for the explanation and good to know that you will be working on it.

Since you said you are going to do some benchmark, do you have a way to tell how much memory LLVM cosume (other than switch versions and disabling certain features?)

**Comment by Keno** [comment]
LLVM does not keep such counts, but you could try one of those heap profiling tools.

**Comment by kmsquire** [comment]
Just saw this (or something like it) in below this error in https://travis-ci.org/JuliaLang/julia/jobs/77843433#L4775-L4779

**Comment by stevengj** [comment]
Can we re-open this?  I'm seeing this reliably in #12855.

**Comment by timholy** [comment]
Any chances of using the "containers" infrastructure (http://docs.travis-ci.com/user/workers/container-based-infrastructure/)? More memory available (4GB rather than 3). Presumably our usage of `sudo` makes this rather hard, however.

**Comment by stevengj** [comment]
Because this is an urgent issue (lots of PRs now have failing builds because of this), my feeling is that we should just set `JULIA_CPU_CORES=2` (as in #12855) for now, and swallow the longer Travis times until there is a better solution.

**Comment by tkelman** [comment]
> Any chances of using the "containers" infrastructure

Not easily, we need to do a lot of apt-get work from the juliadeps PPA which we would need to get whitelisted first, and the set of i386 packages that we install to make 32-bit Linux builds work would likely be a challenge since many of them conflict with the normal 64-bit packages and I don't think Travis lets you do a matrix build with the apt-get addon. In short I haven't tried this and haven't had time to look into it in detail.

**Comment by stevengj** [comment]
Okay, I merged `JULIA_CPU_CORES=2` as a temporary fix.  Slow Travis is better than continual failures.

**Comment by yuyichao** [comment]
Starts happening again.... https://travis-ci.org/JuliaLang/julia/jobs/84749848 :cry:

**Comment by stevengj** [comment]
Re-adding the priority label if we are starting to see continual failures again...  We can set `JULIA_CPU_CORES=1`, of course, but we clearly need a longer-term fix.

**Comment by timholy** [comment]
If JC can spare some cash, would it be worth considering a paid plan? (I note that more memory does **not** seem to be a feature offered in the [plans here](https://travis-ci.com/plans), but I imagine one could ask...) There's been enough time lost to this issue that ponying up might be expedient and cost-effective.

**Comment by StefanKarpinski** [comment]
JC can fund this if Travis offers it and the price is reasonable.

**Comment by stevengj** [comment]
I asked the Travis support people, and got the following response:
>We don't offer paid plans on travis-ci.org (public repositories/builds) at the moment. This is something we are looking into, but due to the requirements this might take a while.
>
>In the meantime, have you considered running these builds on our GCE/Trusty beta infrastructure? There are 7.5 GB of memory available.

**Comment by yuyichao** [comment]
Close since this should be fixed by #13577 now. Add benchmark label since I think we should also have a benchmark to keep track of memory usage. (Feel free to remove the label if there's already one for this purpose.)
